HADOOP-19256. steve's pr of conditional writes

steveloughran opened the pull request:
#7329 with some changes

Commit:
Integrate PutIfNotExist functionality into S3A

Commit:
HADOOP-19256. S3A: Upgrade aws sdk v2 to 2.27.14

Commit:
addressing PR feedback

Commit:
addressing review comments, added flag to PutObjectOptions

Commit:
reverting to using optionheaders

Commit:
reverting to using optionheaders

Commit:
redundant set for multipart_size

Commit:
redundant set for multipart_size

Commit:
fixing merge conflict

Commit:
Fixing tests: testPutIfAbsentConflict, testPutIfAbsentLargeFileConflict

Commit:
Changes based on review comments

steveloughran committed:
HADOOP-19256. review of conditional write

+ hadoop common and s3a create file options

wired up to s3a create file builder, and nowhere else.

Change-Id: Ieaafa9ecdfd68306deb41ecf40c416374a310859

Commit:
HADOOP-19256: add tests

steveloughran committed:
HADOOP-19256. filesystem options for overwrite, etag and small multipart files

createFile() options are passed down all the way to object creation
* fs.option.create.conditional.overwrite: sets the header.
  MUST be in close(). After all, create(overwrite=false) is eager.
* fs.option.create.conditional.overwrite.etag requests etag writes.
  MAY be in create(); may be in close().
* fs.option.create.content.type for mime type.
* fs.option.create.in.close to ask for create in close, as fs capability.
  (maybe make this an fs path capability only?)
* fs.s3a.create.multipart : allows tests to create small multipart files

Javadocs of the fs.option.create try to define semantics; will need
strict fs specification soon.

Bool params come down in an enumset of flags; I'm going to do that in
more code as it is more flexible over time than many booleans.

- etag values are passed down but not wired up/tested
- content type flag exists but is ignored

New WriteObjectFlags enum is in new package o.a.h.fs.s3a.write;
to match the streams package in another ongoing PR.
Goal, as we maintain things, all code related to writing can
go in here.

also: options set by WriteObjectFlags return true in hasCapability()
of the stream

Change-Id: I301abd7397accbd278d05f42f858223ba1349fc8

steveloughran commented:
Tested: S3 london, `-Dparallel-tests -DtestsThreadCount=9`
One failure due to SDK update not telling us off any more.

```


[ERROR] Tests run: 2, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 6.152 s <<< FAILURE! - in org.apache.hadoop.fs.s3a.impl.ITestAwsSdkWorkarounds
[ERROR] testNoisyLogging(org.apache.hadoop.fs.s3a.impl.ITestAwsSdkWorkarounds)  Time elapsed: 2.318 s  <<< FAILURE!
java.lang.AssertionError: 
[LOG output does not contain the forbidden text. Has the SDK been fixed?] 
Expecting:
 <"">
to contain:
 <"The provided S3AsyncClient is an instance of MultipartS3AsyncClient"> 
        at org.apache.hadoop.fs.s3a.impl.ITestAwsSdkWorkarounds.testNoisyLogging(ITestAwsSdkWorkarounds.java:100)
 

+ test timeout of ITestS3APrefetchingLruEviction.testSeeksWithLruEviction

```

steveloughran commented:
@saikatroy038 the two commits here are ready to be cherrypicked.

Commit:
HADOOP-19256: Use forced multipart in tests, add If-Match header in request.

Commit:
HADOOP-19256: Add tests for If-Match

Commit:
HADOOP-19256: Fix javadoc

steveloughran committed:
HADOOP-19256. conditional create.

Another iteration;

* pulled out *all* the stuff related to the notion of a configuration service
  -needs to be independent PR
* "simplified" logic as to when to use conditional create for a normal create()
   call, even without any explicit request.
* doc updates

Change-Id: I6a5770d4d85197f4f9338900cc21f9bca8b7c704

Commit:
HADOOP-19256: Add statistics for conditional writes

Commit:
HADOOP-19256: Add etag in BlockOutputStreamBuilder

Commit:
HADOOP-19256: Add tests for conditional_write statistics

steveloughran committed:
HADOOP-19478. S3A: pull out new configuration load/probes under S3AStore

New service under S3AStoreImpl: StoreConfigurationService

Change-Id: I99c8305574492c05170274dcc363bfba4857981b

steveloughran committed:
HADOOP-19256. Automatic use of conditional create.

S3AFS supports enable/disable of conditional create,
and ability to turn it on automatically.

fs.s3a.conditional.create.enabled:
 option to enable (default) or disable conditional creation

fs.s3a.conditional.create.files:
 option to enable or disable (default) using conditional create when
 writing any file with overwrite = false.

Test setups do not (yet) skip test suites if
fs.s3a.conditional.create.files is false; they should
and it needs to be updated in the third-party stores section

Change-Id: I96a923b84cf29efaaae7ced41bfb72f8944be1ca

hadoop-yetus commented:
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime |  Logfile | Comment |
|:----:|----------:|--------:|:--------:|:-------:|
| +0 :ok: |  reexec  |   7m  2s |  |  Docker mode activated.  |
|||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  |  No case conflicting files found.  |
| +0 :ok: |  codespell  |   0m  0s |  |  codespell was not available.  |
| +0 :ok: |  detsecrets  |   0m  0s |  |  detect-secrets was not available.  |
| +0 :ok: |  xmllint  |   0m  0s |  |  xmllint was not available.  |
| +0 :ok: |  shelldocs  |   0m  0s |  |  Shelldocs was not available.  |
| +1 :green_heart: |  @author  |   0m  0s |  |  The patch does not contain any @author tags.  |
| +1 :green_heart: |  test4tests  |   0m  0s |  |  The patch appears to include 6 new or modified test files.  |
|||| _ trunk Compile Tests _ |
| +0 :ok: |  mvndep  |   6m 14s |  |  Maven dependency ordering for branch  |
| +1 :green_heart: |  mvninstall  |  19m 26s |  |  trunk passed  |
| +1 :green_heart: |  compile  |   9m 10s |  |  trunk passed with JDK Ubuntu-11.0.26+4-post-Ubuntu-1ubuntu120.04  |
| +1 :green_heart: |  compile  |   7m 34s |  |  trunk passed with JDK Private Build-1.8.0_442-8u442-b06~us1-0ubuntu1~20.04-b06  |
| +1 :green_heart: |  checkstyle  |   1m 56s |  |  trunk passed  |
| +1 :green_heart: |  mvnsite  |  14m 48s |  |  trunk passed  |
| -1 :x: |  javadoc  |   1m 31s | [/branch-javadoc-root-jdkUbuntu-11.0.26+4-post-Ubuntu-1ubuntu120.04.txt](https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-7362/1/artifact/out/branch-javadoc-root-jdkUbuntu-11.0.26+4-post-Ubuntu-1ubuntu120.04.txt) |  root in trunk failed with JDK Ubuntu-11.0.26+4-post-Ubuntu-1ubuntu120.04.  |
| +1 :green_heart: |  javadoc  |   4m 42s |  |  trunk passed with JDK Private Build-1.8.0_442-8u442-b06~us1-0ubuntu1~20.04-b06  |
| +0 :ok: |  spotbugs  |   0m 10s |  |  branch/hadoop-project no spotbugs output file (spotbugsXml.xml)  |
| +1 :green_heart: |  shadedclient  |  39m 42s |  |  branch has no errors when building and testing our client artifacts.  |
|||| _ Patch Compile Tests _ |
| +0 :ok: |  mvndep  |   0m 38s |  |  Maven dependency ordering for patch  |
| -1 :x: |  mvninstall  |   0m 15s | [/patch-mvninstall-hadoop-tools_hadoop-aws.txt](https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-7362/1/artifact/out/patch-mvninstall-hadoop-tools_hadoop-aws.txt) |  hadoop-aws in the patch failed.  |
| -1 :x: |  mvninstall  |  21m  2s | [/patch-mvninstall-root.txt](https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-7362/1/artifact/out/patch-mvninstall-root.txt) |  root in the patch failed.  |
| -1 :x: |  compile  |   8m 49s | [/patch-compile-root-jdkUbuntu-11.0.26+4-post-Ubuntu-1ubuntu120.04.txt](https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-7362/1/artifact/out/patch-compile-root-jdkUbuntu-11.0.26+4-post-Ubuntu-1ubuntu120.04.txt) |  root in the patch failed with JDK Ubuntu-11.0.26+4-post-Ubuntu-1ubuntu120.04.  |
| -1 :x: |  javac  |   8m 49s | [/patch-compile-root-jdkUbuntu-11.0.26+4-post-Ubuntu-1ubuntu120.04.txt](https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-7362/1/artifact/out/patch-compile-root-jdkUbuntu-11.0.26+4-post-Ubuntu-1ubuntu120.04.txt) |  root in the patch failed with JDK Ubuntu-11.0.26+4-post-Ubuntu-1ubuntu120.04.  |
| -1 :x: |  compile  |   7m 47s | [/patch-compile-root-jdkPrivateBuild-1.8.0_442-8u442-b06~us1-0ubuntu1~20.04-b06.txt](https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-7362/1/artifact/out/patch-compile-root-jdkPrivateBuild-1.8.0_442-8u442-b06~us1-0ubuntu1~20.04-b06.txt) |  root in the patch failed with JDK Private Build-1.8.0_442-8u442-b06~us1-0ubuntu1~20.04-b06.  |
| -1 :x: |  javac  |   7m 47s | [/patch-compile-root-jdkPrivateBuild-1.8.0_442-8u442-b06~us1-0ubuntu1~20.04-b06.txt](https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-7362/1/artifact/out/patch-compile-root-jdkPrivateBuild-1.8.0_442-8u442-b06~us1-0ubuntu1~20.04-b06.txt) |  root in the patch failed with JDK Private Build-1.8.0_442-8u442-b06~us1-0ubuntu1~20.04-b06.  |
| +1 :green_heart: |  blanks  |   0m  0s |  |  The patch has no blanks issues.  |
| -0 :warning: |  checkstyle  |   2m 12s | [/results-checkstyle-root.txt](https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-7362/1/artifact/out/results-checkstyle-root.txt) |  root: The patch generated 146 new + 23 unchanged - 0 fixed = 169 total (was 23)  |
| -1 :x: |  mvnsite  |   3m 16s | [/patch-mvnsite-root.txt](https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-7362/1/artifact/out/patch-mvnsite-root.txt) |  root in the patch failed.  |
| +1 :green_heart: |  shellcheck  |   0m  0s |  |  No new issues.  |
| -1 :x: |  javadoc  |   0m 50s | [/patch-javadoc-root-jdkUbuntu-11.0.26+4-post-Ubuntu-1ubuntu120.04.txt](https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-7362/1/artifact/out/patch-javadoc-root-jdkUbuntu-11.0.26+4-post-Ubuntu-1ubuntu120.04.txt) |  root in the patch failed with JDK Ubuntu-11.0.26+4-post-Ubuntu-1ubuntu120.04.  |
| -1 :x: |  javadoc  |   4m 40s | [/patch-javadoc-root-jdkPrivateBuild-1.8.0_442-8u442-b06~us1-0ubuntu1~20.04-b06.txt](https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-7362/1/artifact/out/patch-javadoc-root-jdkPrivateBuild-1.8.0_442-8u442-b06~us1-0ubuntu1~20.04-b06.txt) |  root in the patch failed with JDK Private Build-1.8.0_442-8u442-b06~us1-0ubuntu1~20.04-b06.  |
| +0 :ok: |  spotbugs  |   0m 13s |  |  hadoop-project has no data from spotbugs  |
| -1 :x: |  spotbugs  |   0m 14s | [/patch-spotbugs-hadoop-tools_hadoop-aws.txt](https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-7362/1/artifact/out/patch-spotbugs-hadoop-tools_hadoop-aws.txt) |  hadoop-aws in the patch failed.  |
| -1 :x: |  spotbugs  |  17m 47s | [/patch-spotbugs-root.txt](https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-7362/1/artifact/out/patch-spotbugs-root.txt) |  root in the patch failed.  |
| +1 :green_heart: |  shadedclient  |  24m 47s |  |  patch has no errors when building and testing our client artifacts.  |
|||| _ Other Tests _ |
| -1 :x: |  unit  | 472m 29s | [/patch-unit-root.txt](https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-7362/1/artifact/out/patch-unit-root.txt) |  root in the patch failed.  |
| +1 :green_heart: |  asflicense  |   0m 51s |  |  The patch does not generate ASF License warnings.  |
|  |   | 684m 31s |  |  |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.48 ServerAPI=1.48 base: https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-7362/1/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/7362 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient spotbugs checkstyle codespell detsecrets xmllint shellcheck shelldocs |
| uname | Linux 771a59a22012 5.15.0-130-generic #140-Ubuntu SMP Wed Dec 18 17:59:53 UTC 2024 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev-support/bin/hadoop.sh |
| git revision | trunk / fe59fc0872e00a9e7fa6b9e95d884c2ff9bd3a57 |
| Default Java | Private Build-1.8.0_442-8u442-b06~us1-0ubuntu1~20.04-b06 |
| Multi-JDK versions | /usr/lib/jvm/java-11-openjdk-amd64:Ubuntu-11.0.26+4-post-Ubuntu-1ubuntu120.04 /usr/lib/jvm/java-8-openjdk-amd64:Private Build-1.8.0_442-8u442-b06~us1-0ubuntu1~20.04-b06 |
|  Test Results | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-7362/1/testReport/ |
| Max. process+thread count | 3151 (vs. ulimit of 5500) |
| modules | C: hadoop-project hadoop-common-project/hadoop-common hadoop-tools/hadoop-aws . U: . |
| Console output | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-7362/1/console |
| versions | git=2.25.1 maven=3.6.3 spotbugs=4.2.2 shellcheck=0.7.0 |
| Powered by | Apache Yetus 0.14.0 https://yetus.apache.org |


This message was automatically generated.

hadoop-yetus commented:
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime |  Logfile | Comment |
|:----:|----------:|--------:|:--------:|:-------:|
| +0 :ok: |  reexec  |   0m 21s |  |  Docker mode activated.  |
|||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  1s |  |  No case conflicting files found.  |
| +0 :ok: |  codespell  |   0m  0s |  |  codespell was not available.  |
| +0 :ok: |  detsecrets  |   0m  0s |  |  detect-secrets was not available.  |
| +0 :ok: |  xmllint  |   0m  0s |  |  xmllint was not available.  |
| +0 :ok: |  shelldocs  |   0m  0s |  |  Shelldocs was not available.  |
| +1 :green_heart: |  @author  |   0m  0s |  |  The patch does not contain any @author tags.  |
| +1 :green_heart: |  test4tests  |   0m  0s |  |  The patch appears to include 7 new or modified test files.  |
|||| _ trunk Compile Tests _ |
| +0 :ok: |  mvndep  |   6m 41s |  |  Maven dependency ordering for branch  |
| +1 :green_heart: |  mvninstall  |  24m 14s |  |  trunk passed  |
| +1 :green_heart: |  compile  |  10m  5s |  |  trunk passed with JDK Ubuntu-11.0.26+4-post-Ubuntu-1ubuntu120.04  |
| +1 :green_heart: |  compile  |   9m  3s |  |  trunk passed with JDK Private Build-1.8.0_442-8u442-b06~us1-0ubuntu1~20.04-b06  |
| +1 :green_heart: |  checkstyle  |   2m 36s |  |  trunk passed  |
| +1 :green_heart: |  mvnsite  |  14m 55s |  |  trunk passed  |
| -1 :x: |  javadoc  |   1m 31s | [/branch-javadoc-root-jdkUbuntu-11.0.26+4-post-Ubuntu-1ubuntu120.04.txt](https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-7362/2/artifact/out/branch-javadoc-root-jdkUbuntu-11.0.26+4-post-Ubuntu-1ubuntu120.04.txt) |  root in trunk failed with JDK Ubuntu-11.0.26+4-post-Ubuntu-1ubuntu120.04.  |
| +1 :green_heart: |  javadoc  |   4m 41s |  |  trunk passed with JDK Private Build-1.8.0_442-8u442-b06~us1-0ubuntu1~20.04-b06  |
| +0 :ok: |  spotbugs  |   0m 11s |  |  branch/hadoop-project no spotbugs output file (spotbugsXml.xml)  |
| +1 :green_heart: |  shadedclient  |  43m 10s |  |  branch has no errors when building and testing our client artifacts.  |
|||| _ Patch Compile Tests _ |
| +0 :ok: |  mvndep  |   0m 33s |  |  Maven dependency ordering for patch  |
| +1 :green_heart: |  mvninstall  |  20m 37s |  |  the patch passed  |
| +1 :green_heart: |  compile  |   9m 21s |  |  the patch passed with JDK Ubuntu-11.0.26+4-post-Ubuntu-1ubuntu120.04  |
| +1 :green_heart: |  javac  |   9m 21s |  |  the patch passed  |
| +1 :green_heart: |  compile  |   8m 46s |  |  the patch passed with JDK Private Build-1.8.0_442-8u442-b06~us1-0ubuntu1~20.04-b06  |
| +1 :green_heart: |  javac  |   8m 46s |  |  the patch passed  |
| +1 :green_heart: |  blanks  |   0m  0s |  |  The patch has no blanks issues.  |
| -0 :warning: |  checkstyle  |   1m 59s | [/results-checkstyle-root.txt](https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-7362/2/artifact/out/results-checkstyle-root.txt) |  root: The patch generated 331 new + 23 unchanged - 0 fixed = 354 total (was 23)  |
| +1 :green_heart: |  mvnsite  |  12m 45s |  |  the patch passed  |
| +1 :green_heart: |  shellcheck  |   0m  0s |  |  No new issues.  |
| -1 :x: |  javadoc  |   0m 53s | [/patch-javadoc-root-jdkUbuntu-11.0.26+4-post-Ubuntu-1ubuntu120.04.txt](https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-7362/2/artifact/out/patch-javadoc-root-jdkUbuntu-11.0.26+4-post-Ubuntu-1ubuntu120.04.txt) |  root in the patch failed with JDK Ubuntu-11.0.26+4-post-Ubuntu-1ubuntu120.04.  |
| -1 :x: |  javadoc  |   4m 48s | [/patch-javadoc-root-jdkPrivateBuild-1.8.0_442-8u442-b06~us1-0ubuntu1~20.04-b06.txt](https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-7362/2/artifact/out/patch-javadoc-root-jdkPrivateBuild-1.8.0_442-8u442-b06~us1-0ubuntu1~20.04-b06.txt) |  root in the patch failed with JDK Private Build-1.8.0_442-8u442-b06~us1-0ubuntu1~20.04-b06.  |
| +0 :ok: |  spotbugs  |   0m 13s |  |  hadoop-project has no data from spotbugs  |
| -1 :x: |  spotbugs  |   0m 50s | [/new-spotbugs-hadoop-tools_hadoop-aws.html](https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-7362/2/artifact/out/new-spotbugs-hadoop-tools_hadoop-aws.html) |  hadoop-tools/hadoop-aws generated 2 new + 0 unchanged - 0 fixed = 2 total (was 0)  |
| -1 :x: |  spotbugs  |  19m 43s | [/new-spotbugs-root.html](https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-7362/2/artifact/out/new-spotbugs-root.html) |  root generated 2 new + 0 unchanged - 0 fixed = 2 total (was 0)  |
| +1 :green_heart: |  shadedclient  |  42m  5s |  |  patch has no errors when building and testing our client artifacts.  |
|||| _ Other Tests _ |
| +1 :green_heart: |  unit  | 477m 57s |  |  root in the patch passed.  |
| +1 :green_heart: |  asflicense  |   0m 57s |  |  The patch does not generate ASF License warnings.  |
|  |   | 706m 10s |  |  |


| Reason | Tests |
|-------:|:------|
| SpotBugs | module:hadoop-tools/hadoop-aws |
|  |  Unread field:S3ABlockOutputStream.java:[line 1607] |
|  |  Unread field:S3ABlockOutputStream.java:[line 1601] |
| SpotBugs | module:root |
|  |  Unread field:S3ABlockOutputStream.java:[line 1607] |
|  |  Unread field:S3ABlockOutputStream.java:[line 1601] |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.48 ServerAPI=1.48 base: https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-7362/2/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/7362 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient spotbugs checkstyle codespell detsecrets xmllint shellcheck shelldocs |
| uname | Linux 88aac3106aab 5.15.0-130-generic #140-Ubuntu SMP Wed Dec 18 17:59:53 UTC 2024 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev-support/bin/hadoop.sh |
| git revision | trunk / 6a59bb13d96aba20e5a4c60dd9a4a962b520cda5 |
| Default Java | Private Build-1.8.0_442-8u442-b06~us1-0ubuntu1~20.04-b06 |
| Multi-JDK versions | /usr/lib/jvm/java-11-openjdk-amd64:Ubuntu-11.0.26+4-post-Ubuntu-1ubuntu120.04 /usr/lib/jvm/java-8-openjdk-amd64:Private Build-1.8.0_442-8u442-b06~us1-0ubuntu1~20.04-b06 |
|  Test Results | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-7362/2/testReport/ |
| Max. process+thread count | 3241 (vs. ulimit of 5500) |
| modules | C: hadoop-project hadoop-common-project/hadoop-common hadoop-tools/hadoop-aws . U: . |
| Console output | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-7362/2/console |
| versions | git=2.25.1 maven=3.6.3 spotbugs=4.2.2 shellcheck=0.7.0 |
| Powered by | Apache Yetus 0.14.0 https://yetus.apache.org |


This message was automatically generated.

Commit:
HADOOP-19256: Changes as per review comments

Commit:
HADOOP-19256: Add tests

Commit:
HADOOP-19256: Update tests

Commit:
HADOOP-19256: Remove incorrect tests.

- testIfMatchCreateFileWithoutOverwrite, testIfMatchCreateFileWithoutOverwriteWithPerformanceFlag
- with the new design, overwrite is ignored if etag is passed

Commit:
HADOOP-19256: Update tests, fix spotBugs

Commit:
HADOOP-19256: Ignore condition_write statistics test

saikatroy038 committed:
HADOOP-19256: Merge branch 'trunk' into s3-conditional-writes

steveloughran committed:
HADOOP-19256. Final style checks for yetus

steveloughran commented:
Forced push of saikat's final PR and a checkstyle fix. final test run and review underway...If I'm happy this will be what I merge in

steveloughran committed:
HADOOP-19256. Nits from code review

steveloughran commented:
I am +1 pending yetus; while I've done some of this most of the work is from others

hadoop-yetus commented:
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime |  Logfile | Comment |
|:----:|----------:|--------:|:--------:|:-------:|
| +0 :ok: |  reexec  |   1m  0s |  |  Docker mode activated.  |
|||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  |  No case conflicting files found.  |
| +0 :ok: |  codespell  |   0m  1s |  |  codespell was not available.  |
| +0 :ok: |  detsecrets  |   0m  1s |  |  detect-secrets was not available.  |
| +0 :ok: |  markdownlint  |   0m  1s |  |  markdownlint was not available.  |
| +1 :green_heart: |  @author  |   0m  0s |  |  The patch does not contain any @author tags.  |
| +1 :green_heart: |  test4tests  |   0m  0s |  |  The patch appears to include 9 new or modified test files.  |
|||| _ trunk Compile Tests _ |
| +0 :ok: |  mvndep  |   6m 17s |  |  Maven dependency ordering for branch  |
| +1 :green_heart: |  mvninstall  |  31m 32s |  |  trunk passed  |
| +1 :green_heart: |  compile  |  15m 33s |  |  trunk passed with JDK Ubuntu-11.0.26+4-post-Ubuntu-1ubuntu120.04  |
| +1 :green_heart: |  compile  |  13m 27s |  |  trunk passed with JDK Private Build-1.8.0_442-8u442-b06~us1-0ubuntu1~20.04-b06  |
| +1 :green_heart: |  checkstyle  |   4m 28s |  |  trunk passed  |
| +1 :green_heart: |  mvnsite  |   2m 32s |  |  trunk passed  |
| +1 :green_heart: |  javadoc  |   2m  8s |  |  trunk passed with JDK Ubuntu-11.0.26+4-post-Ubuntu-1ubuntu120.04  |
| +1 :green_heart: |  javadoc  |   1m 45s |  |  trunk passed with JDK Private Build-1.8.0_442-8u442-b06~us1-0ubuntu1~20.04-b06  |
| +1 :green_heart: |  spotbugs  |   3m 53s |  |  trunk passed  |
| +1 :green_heart: |  shadedclient  |  34m 51s |  |  branch has no errors when building and testing our client artifacts.  |
| -0 :warning: |  patch  |  35m 17s |  |  Used diff version of patch file. Binary files and potentially other changes not applied. Please rebase and squash commits if necessary.  |
|||| _ Patch Compile Tests _ |
| +0 :ok: |  mvndep  |   0m 33s |  |  Maven dependency ordering for patch  |
| +1 :green_heart: |  mvninstall  |   1m 26s |  |  the patch passed  |
| +1 :green_heart: |  compile  |  14m 52s |  |  the patch passed with JDK Ubuntu-11.0.26+4-post-Ubuntu-1ubuntu120.04  |
| +1 :green_heart: |  javac  |  14m 52s |  |  the patch passed  |
| +1 :green_heart: |  compile  |  13m 43s |  |  the patch passed with JDK Private Build-1.8.0_442-8u442-b06~us1-0ubuntu1~20.04-b06  |
| +1 :green_heart: |  javac  |  13m 43s |  |  the patch passed  |
| +1 :green_heart: |  blanks  |   0m  0s |  |  The patch has no blanks issues.  |
| -0 :warning: |  checkstyle  |   4m 14s | [/results-checkstyle-root.txt](https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-7362/4/artifact/out/results-checkstyle-root.txt) |  root: The patch generated 3 new + 23 unchanged - 0 fixed = 26 total (was 23)  |
| +1 :green_heart: |  mvnsite  |   2m 34s |  |  the patch passed  |
| +1 :green_heart: |  javadoc  |   2m  3s |  |  the patch passed with JDK Ubuntu-11.0.26+4-post-Ubuntu-1ubuntu120.04  |
| +1 :green_heart: |  javadoc  |   1m 41s |  |  the patch passed with JDK Private Build-1.8.0_442-8u442-b06~us1-0ubuntu1~20.04-b06  |
| -1 :x: |  spotbugs  |   2m 42s | [/new-spotbugs-hadoop-common-project_hadoop-common.html](https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-7362/4/artifact/out/new-spotbugs-hadoop-common-project_hadoop-common.html) |  hadoop-common-project/hadoop-common generated 1 new + 0 unchanged - 0 fixed = 1 total (was 0)  |
| +1 :green_heart: |  shadedclient  |  34m 56s |  |  patch has no errors when building and testing our client artifacts.  |
|||| _ Other Tests _ |
| +1 :green_heart: |  unit  |  15m  2s |  |  hadoop-common in the patch passed.  |
| +1 :green_heart: |  unit  |   3m 40s |  |  hadoop-aws in the patch passed.  |
| +1 :green_heart: |  asflicense  |   1m  4s |  |  The patch does not generate ASF License warnings.  |
|  |   | 221m 51s |  |  |


| Reason | Tests |
|-------:|:------|
| SpotBugs | module:hadoop-common-project/hadoop-common |
|  |  Should org.apache.hadoop.fs.Options$CreateFileOptionKeys be a _static_ inner class?  At Options.java:inner class?  At Options.java:[line 723] |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.48 ServerAPI=1.48 base: https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-7362/4/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/7362 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient spotbugs checkstyle codespell detsecrets markdownlint |
| uname | Linux 16b31b124f3f 5.15.0-130-generic #140-Ubuntu SMP Wed Dec 18 17:59:53 UTC 2024 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev-support/bin/hadoop.sh |
| git revision | trunk / d1a82eccb71ecc3539404352f59154058ed6a65f |
| Default Java | Private Build-1.8.0_442-8u442-b06~us1-0ubuntu1~20.04-b06 |
| Multi-JDK versions | /usr/lib/jvm/java-11-openjdk-amd64:Ubuntu-11.0.26+4-post-Ubuntu-1ubuntu120.04 /usr/lib/jvm/java-8-openjdk-amd64:Private Build-1.8.0_442-8u442-b06~us1-0ubuntu1~20.04-b06 |
|  Test Results | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-7362/4/testReport/ |
| Max. process+thread count | 3151 (vs. ulimit of 5500) |
| modules | C: hadoop-common-project/hadoop-common hadoop-tools/hadoop-aws U: . |
| Console output | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-7362/4/console |
| versions | git=2.25.1 maven=3.6.3 spotbugs=4.2.2 |
| Powered by | Apache Yetus 0.14.0 https://yetus.apache.org |


This message was automatically generated.

hadoop-yetus commented:
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime |  Logfile | Comment |
|:----:|----------:|--------:|:--------:|:-------:|
| +0 :ok: |  reexec  |   0m 22s |  |  Docker mode activated.  |
|||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  |  No case conflicting files found.  |
| +0 :ok: |  codespell  |   0m  1s |  |  codespell was not available.  |
| +0 :ok: |  detsecrets  |   0m  1s |  |  detect-secrets was not available.  |
| +0 :ok: |  markdownlint  |   0m  1s |  |  markdownlint was not available.  |
| +1 :green_heart: |  @author  |   0m  0s |  |  The patch does not contain any @author tags.  |
| +1 :green_heart: |  test4tests  |   0m  0s |  |  The patch appears to include 9 new or modified test files.  |
|||| _ trunk Compile Tests _ |
| +0 :ok: |  mvndep  |   5m 47s |  |  Maven dependency ordering for branch  |
| +1 :green_heart: |  mvninstall  |  22m 18s |  |  trunk passed  |
| +1 :green_heart: |  compile  |   9m 16s |  |  trunk passed with JDK Ubuntu-11.0.26+4-post-Ubuntu-1ubuntu120.04  |
| +1 :green_heart: |  compile  |   7m 57s |  |  trunk passed with JDK Private Build-1.8.0_442-8u442-b06~us1-0ubuntu1~20.04-b06  |
| +1 :green_heart: |  checkstyle  |   2m  6s |  |  trunk passed  |
| +1 :green_heart: |  mvnsite  |   1m 40s |  |  trunk passed  |
| +1 :green_heart: |  javadoc  |   1m  6s |  |  trunk passed with JDK Ubuntu-11.0.26+4-post-Ubuntu-1ubuntu120.04  |
| +1 :green_heart: |  javadoc  |   0m 52s |  |  trunk passed with JDK Private Build-1.8.0_442-8u442-b06~us1-0ubuntu1~20.04-b06  |
| +1 :green_heart: |  spotbugs  |   2m 17s |  |  trunk passed  |
| +1 :green_heart: |  shadedclient  |  23m 21s |  |  branch has no errors when building and testing our client artifacts.  |
| -0 :warning: |  patch  |  23m 37s |  |  Used diff version of patch file. Binary files and potentially other changes not applied. Please rebase and squash commits if necessary.  |
|||| _ Patch Compile Tests _ |
| +0 :ok: |  mvndep  |   0m 23s |  |  Maven dependency ordering for patch  |
| -1 :x: |  mvninstall  |   0m 14s | [/patch-mvninstall-hadoop-tools_hadoop-aws.txt](https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-7362/5/artifact/out/patch-mvninstall-hadoop-tools_hadoop-aws.txt) |  hadoop-aws in the patch failed.  |
| -1 :x: |  compile  |   8m 45s | [/patch-compile-root-jdkUbuntu-11.0.26+4-post-Ubuntu-1ubuntu120.04.txt](https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-7362/5/artifact/out/patch-compile-root-jdkUbuntu-11.0.26+4-post-Ubuntu-1ubuntu120.04.txt) |  root in the patch failed with JDK Ubuntu-11.0.26+4-post-Ubuntu-1ubuntu120.04.  |
| -1 :x: |  javac  |   8m 45s | [/patch-compile-root-jdkUbuntu-11.0.26+4-post-Ubuntu-1ubuntu120.04.txt](https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-7362/5/artifact/out/patch-compile-root-jdkUbuntu-11.0.26+4-post-Ubuntu-1ubuntu120.04.txt) |  root in the patch failed with JDK Ubuntu-11.0.26+4-post-Ubuntu-1ubuntu120.04.  |
| -1 :x: |  compile  |   7m 53s | [/patch-compile-root-jdkPrivateBuild-1.8.0_442-8u442-b06~us1-0ubuntu1~20.04-b06.txt](https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-7362/5/artifact/out/patch-compile-root-jdkPrivateBuild-1.8.0_442-8u442-b06~us1-0ubuntu1~20.04-b06.txt) |  root in the patch failed with JDK Private Build-1.8.0_442-8u442-b06~us1-0ubuntu1~20.04-b06.  |
| -1 :x: |  javac  |   7m 53s | [/patch-compile-root-jdkPrivateBuild-1.8.0_442-8u442-b06~us1-0ubuntu1~20.04-b06.txt](https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-7362/5/artifact/out/patch-compile-root-jdkPrivateBuild-1.8.0_442-8u442-b06~us1-0ubuntu1~20.04-b06.txt) |  root in the patch failed with JDK Private Build-1.8.0_442-8u442-b06~us1-0ubuntu1~20.04-b06.  |
| +1 :green_heart: |  blanks  |   0m  0s |  |  The patch has no blanks issues.  |
| -0 :warning: |  checkstyle  |   2m 18s | [/results-checkstyle-root.txt](https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-7362/5/artifact/out/results-checkstyle-root.txt) |  root: The patch generated 3 new + 23 unchanged - 0 fixed = 26 total (was 23)  |
| -1 :x: |  mvnsite  |   0m 23s | [/patch-mvnsite-hadoop-tools_hadoop-aws.txt](https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-7362/5/artifact/out/patch-mvnsite-hadoop-tools_hadoop-aws.txt) |  hadoop-aws in the patch failed.  |
| +1 :green_heart: |  javadoc  |   1m  7s |  |  the patch passed with JDK Ubuntu-11.0.26+4-post-Ubuntu-1ubuntu120.04  |
| +1 :green_heart: |  javadoc  |   1m  0s |  |  the patch passed with JDK Private Build-1.8.0_442-8u442-b06~us1-0ubuntu1~20.04-b06  |
| -1 :x: |  spotbugs  |   1m 42s | [/new-spotbugs-hadoop-common-project_hadoop-common.html](https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-7362/5/artifact/out/new-spotbugs-hadoop-common-project_hadoop-common.html) |  hadoop-common-project/hadoop-common generated 1 new + 0 unchanged - 0 fixed = 1 total (was 0)  |
| -1 :x: |  spotbugs  |   0m 24s | [/patch-spotbugs-hadoop-tools_hadoop-aws.txt](https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-7362/5/artifact/out/patch-spotbugs-hadoop-tools_hadoop-aws.txt) |  hadoop-aws in the patch failed.  |
| +1 :green_heart: |  shadedclient  |  25m 54s |  |  patch has no errors when building and testing our client artifacts.  |
|||| _ Other Tests _ |
| +1 :green_heart: |  unit  |  12m 16s |  |  hadoop-common in the patch passed.  |
| -1 :x: |  unit  |   0m 25s | [/patch-unit-hadoop-tools_hadoop-aws.txt](https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-7362/5/artifact/out/patch-unit-hadoop-tools_hadoop-aws.txt) |  hadoop-aws in the patch failed.  |
| +1 :green_heart: |  asflicense  |   0m 31s |  |  The patch does not generate ASF License warnings.  |
|  |   | 143m 28s |  |  |


| Reason | Tests |
|-------:|:------|
| SpotBugs | module:hadoop-common-project/hadoop-common |
|  |  Should org.apache.hadoop.fs.Options$CreateFileOptionKeys be a _static_ inner class?  At Options.java:inner class?  At Options.java:[line 723] |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.48 ServerAPI=1.48 base: https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-7362/5/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/7362 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient spotbugs checkstyle codespell detsecrets markdownlint |
| uname | Linux 9b560b8ce3db 5.15.0-130-generic #140-Ubuntu SMP Wed Dec 18 17:59:53 UTC 2024 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev-support/bin/hadoop.sh |
| git revision | trunk / cac4470b0b8b789806458cbea9b47bd905c84b87 |
| Default Java | Private Build-1.8.0_442-8u442-b06~us1-0ubuntu1~20.04-b06 |
| Multi-JDK versions | /usr/lib/jvm/java-11-openjdk-amd64:Ubuntu-11.0.26+4-post-Ubuntu-1ubuntu120.04 /usr/lib/jvm/java-8-openjdk-amd64:Private Build-1.8.0_442-8u442-b06~us1-0ubuntu1~20.04-b06 |
|  Test Results | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-7362/5/testReport/ |
| Max. process+thread count | 3150 (vs. ulimit of 5500) |
| modules | C: hadoop-common-project/hadoop-common hadoop-tools/hadoop-aws U: . |
| Console output | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-7362/5/console |
| versions | git=2.25.1 maven=3.6.3 spotbugs=4.2.2 |
| Powered by | Apache Yetus 0.14.0 https://yetus.apache.org |


This message was automatically generated.

steveloughran committed:
HADOOP-19256. fix accidental edit; style

hadoop-yetus commented:
:confetti_ball: **+1 overall**






| Vote | Subsystem | Runtime |  Logfile | Comment |
|:----:|----------:|--------:|:--------:|:-------:|
| +0 :ok: |  reexec  |   0m 57s |  |  Docker mode activated.  |
|||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  |  No case conflicting files found.  |
| +0 :ok: |  codespell  |   0m  1s |  |  codespell was not available.  |
| +0 :ok: |  detsecrets  |   0m  1s |  |  detect-secrets was not available.  |
| +0 :ok: |  markdownlint  |   0m  1s |  |  markdownlint was not available.  |
| +1 :green_heart: |  @author  |   0m  0s |  |  The patch does not contain any @author tags.  |
| +1 :green_heart: |  test4tests  |   0m  0s |  |  The patch appears to include 9 new or modified test files.  |
|||| _ trunk Compile Tests _ |
| +0 :ok: |  mvndep  |   6m  6s |  |  Maven dependency ordering for branch  |
| +1 :green_heart: |  mvninstall  |  31m 47s |  |  trunk passed  |
| +1 :green_heart: |  compile  |  15m 36s |  |  trunk passed with JDK Ubuntu-11.0.26+4-post-Ubuntu-1ubuntu120.04  |
| +1 :green_heart: |  compile  |  13m 34s |  |  trunk passed with JDK Private Build-1.8.0_442-8u442-b06~us1-0ubuntu1~20.04-b06  |
| +1 :green_heart: |  checkstyle  |   4m 12s |  |  trunk passed  |
| +1 :green_heart: |  mvnsite  |   2m 35s |  |  trunk passed  |
| +1 :green_heart: |  javadoc  |   2m 10s |  |  trunk passed with JDK Ubuntu-11.0.26+4-post-Ubuntu-1ubuntu120.04  |
| +1 :green_heart: |  javadoc  |   1m 42s |  |  trunk passed with JDK Private Build-1.8.0_442-8u442-b06~us1-0ubuntu1~20.04-b06  |
| +1 :green_heart: |  spotbugs  |   3m 51s |  |  trunk passed  |
| +1 :green_heart: |  shadedclient  |  34m 51s |  |  branch has no errors when building and testing our client artifacts.  |
| -0 :warning: |  patch  |  35m 17s |  |  Used diff version of patch file. Binary files and potentially other changes not applied. Please rebase and squash commits if necessary.  |
|||| _ Patch Compile Tests _ |
| +0 :ok: |  mvndep  |   0m 32s |  |  Maven dependency ordering for patch  |
| +1 :green_heart: |  mvninstall  |   1m 25s |  |  the patch passed  |
| +1 :green_heart: |  compile  |  14m 54s |  |  the patch passed with JDK Ubuntu-11.0.26+4-post-Ubuntu-1ubuntu120.04  |
| +1 :green_heart: |  javac  |  14m 54s |  |  the patch passed  |
| +1 :green_heart: |  compile  |  13m 36s |  |  the patch passed with JDK Private Build-1.8.0_442-8u442-b06~us1-0ubuntu1~20.04-b06  |
| +1 :green_heart: |  javac  |  13m 36s |  |  the patch passed  |
| +1 :green_heart: |  blanks  |   0m  0s |  |  The patch has no blanks issues.  |
| +1 :green_heart: |  checkstyle  |   4m  8s |  |  the patch passed  |
| +1 :green_heart: |  mvnsite  |   2m 31s |  |  the patch passed  |
| +1 :green_heart: |  javadoc  |   2m  4s |  |  the patch passed with JDK Ubuntu-11.0.26+4-post-Ubuntu-1ubuntu120.04  |
| +1 :green_heart: |  javadoc  |   1m 44s |  |  the patch passed with JDK Private Build-1.8.0_442-8u442-b06~us1-0ubuntu1~20.04-b06  |
| +1 :green_heart: |  spotbugs  |   4m 12s |  |  the patch passed  |
| +1 :green_heart: |  shadedclient  |  34m 49s |  |  patch has no errors when building and testing our client artifacts.  |
|||| _ Other Tests _ |
| +1 :green_heart: |  unit  |  15m  4s |  |  hadoop-common in the patch passed.  |
| +1 :green_heart: |  unit  |   3m 40s |  |  hadoop-aws in the patch passed.  |
| +1 :green_heart: |  asflicense  |   1m  7s |  |  The patch does not generate ASF License warnings.  |
|  |   | 221m 28s |  |  |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.48 ServerAPI=1.48 base: https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-7362/6/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/7362 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient spotbugs checkstyle codespell detsecrets markdownlint |
| uname | Linux 5e46cf543208 5.15.0-130-generic #140-Ubuntu SMP Wed Dec 18 17:59:53 UTC 2024 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev-support/bin/hadoop.sh |
| git revision | trunk / 6a511c9b2cf019d8c739b598dc14cf163344f171 |
| Default Java | Private Build-1.8.0_442-8u442-b06~us1-0ubuntu1~20.04-b06 |
| Multi-JDK versions | /usr/lib/jvm/java-11-openjdk-amd64:Ubuntu-11.0.26+4-post-Ubuntu-1ubuntu120.04 /usr/lib/jvm/java-8-openjdk-amd64:Private Build-1.8.0_442-8u442-b06~us1-0ubuntu1~20.04-b06 |
|  Test Results | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-7362/6/testReport/ |
| Max. process+thread count | 1261 (vs. ulimit of 5500) |
| modules | C: hadoop-common-project/hadoop-common hadoop-tools/hadoop-aws U: . |
| Console output | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-7362/6/console |
| versions | git=2.25.1 maven=3.6.3 spotbugs=4.2.2 |
| Powered by | Apache Yetus 0.14.0 https://yetus.apache.org |


This message was automatically generated.

steveloughran commented:
ok, ready to merge. doing a local flattening of the PR as with the merges and things it's got complex. then will push up the final single patch
